### PR TITLE
Swenzel/tpcimpr

### PR DESCRIPTION
**Pull request by sawenzel**
Improvements for TPC:

  * sectorwise hit storage (with a runtime gain in memory usage)
  * sectorwise hit processing in digitization
  * be able to switch on/off FairLinks during compilation for TPC digits
  

General improvements:

  * An event generator being able to read from an (AliRoot) kinematics file

### CODE REVIEW

File: Detectors/TPC/simulation/include/TPCSimulation/Point.h [hunk 1]
```diff
@@ -34,24 +45,139 @@ class ElementalHit {
...
-  LinkableHitGroup(int trackID) : mHits() {
+  LinkableHitGroup() :
+#ifdef HIT_AOS
+  mHits()
+#else
+  mHitsX_v(),
+  mHitsY_v(),
+  mHitsZ_v(),
+  mHitsT_v(),
+  mHitsE_v(),
```
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/Point.h]: Aren't underscores against the camel case convention?

File: Detectors/TPC/simulation/include/TPCSimulation/Point.h [hunk 2]
```diff
@@ -34,24 +45,139 @@ class ElementalHit {
...
+
+  LinkableHitGroup(int trackID) :
+#ifdef HIT_AOS
+  mHits()
+#else
+  mHitsX_v(),
+  mHitsY_v(),
+  mHitsZ_v(),
+  mHitsT_v(),
+  mHitsE_v(),
```
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/Point.h]: Same here

File: Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h [hunk 3]
```diff
@@ -25,7 +25,7 @@ class DigitizerTask : public FairTask{
   public:
       
     /// Default constructor
-    DigitizerTask();
+    DigitizerTask(int s=-1);
```
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h]: if `s` is sector, perhaps call it `sector`

File: Detectors/TPC/simulation/include/TPCSimulation/Point.h [hunk 4]
```diff
@@ -34,24 +45,139 @@ class ElementalHit {
...
+    mHitsY_v.emplace_back(y);
+    mHitsZ_v.emplace_back(z);
+    mHitsT_v.emplace_back(time);
+    mHitsE_v.emplace_back(e);
+    mSize=mHitsX_v.size();
+    mHitsX=&mHitsX_v[0];
+    mHitsY=&mHitsY_v[0];
+    mHitsZ=&mHitsZ_v[0];
+    mHitsT=&mHitsT_v[0];
+    mHitsE=&mHitsE_v[0];
```
Comment (wiechula) [inline Detectors/TPC/simulation/include/TPCSimulation/Point.h]: What's the reason to use single vectors for each variable?
What it HIT_AOS?

Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/Point.h]: single vectors for structure of arrays storage. This leads to a "manual" splitting of the container inside the TClonesArray and to a 10% better compression in the ROOT file. (Unfortunately, ROOT is not able to split a std::vector<Hits> inside a TClonesArray itself.

The HIT_AOS option is just to go back to the array of structure storage.

File: Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h (hunk 3 above)
Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h]: will do

File: Detectors/TPC/simulation/include/TPCSimulation/Point.h (hunk 2 above)
Comment (sawenzel) [inline Detectors/TPC/simulation/include/TPCSimulation/Point.h]: will change

File: Generators/include/Generators/GeneratorFromFile.h [hunk 5]
```diff
@@ -0,0 +1,45 @@
+/// \author S. Wenzel - Mai 2017
+
+#ifndef ALICEO2_GEN_FROMKINEMATICS_H_
```
Comment (matthiasrichter) [inline Generators/include/Generators/GeneratorFromFile.h]: `O2_EVENTGEN_GENERATORFROMFILE_H_` to follow namespaces and file name?

File: Generators/src/GeneratorFromFile.cxx [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+      if (object != nullptr) mEventsAvailable++;
+    }
+    while (object != nullptr);
+    std::cout << "Found " << mEventsAvailable << " events in this file \n";
+  }
+  
+  Bool_t GeneratorFromFile::ReadEvent(FairPrimaryGenerator* primGen) {
+    if (mEventCounter < mEventsAvailable) {
+	// get the tree and the branch
+	char treestring[14];
```
Comment (matthiasrichter) [inline Generators/src/GeneratorFromFile.cxx]: that limits the event count to 99, then the string is longer than the field. Use `stringstream` instead?

File: Generators/src/GeneratorFromFile.cxx [hunk 7]
```diff
@@ -0,0 +1,78 @@
...
+	sprintf(treestring, "Event%d/TreeK", mEventCounter);
+        TTree *tree = (TTree*) mEventFile->Get(treestring);
+	if (tree == nullptr) return kFALSE;
+
+	auto branch = tree->GetBranch("Particles");
+        TParticle *primary = new TParticle();
```
Comment (matthiasrichter) [inline Generators/src/GeneratorFromFile.cxx]: make line indent consistent in the lines starting from here

File: cmake/O2Dependencies.cmake [hunk 8]
```diff
@@ -639,7 +640,8 @@ o2_define_bucket(
     generators_bucket
 
     DEPENDENCIES
-    Base SimulationDataFormat pythia6 pythia8 MathCore
+    #   Base SimulationDataFormat pythia6 pythia8 MathCore RIO Tree
```
Comment (matthiasrichter) [inline cmake/O2Dependencies.cmake]: is this the idea to remove pythia generators completely? Then the `Generators/CMakelists.txt` has to be revised.

Comment (sawenzel) [inline cmake/O2Dependencies.cmake]: no: this is just temporarily. The old setup compiled but had runtime problems. I will create an issue for this to re-enable them.

**Comment by sawenzel** [comment]
@wiechula  + @matthiasrichter : Your suggestions have been addressed.

File: Generators/src/GeneratorFromFile.cxx [hunk 9]
```diff
@@ -0,0 +1,89 @@
...
+  mEventFile = TFile::Open(name);
+  if (mEventFile == nullptr) {
+    LOG(ERROR) << "EventFile " << name << " not found \n";
+    return;
+  }
+  // the kinematics will be stored inside a Tree "TreeK" with branch "Particles"
+  // different events are stored inside TDirectories
+
+  // we need to probe for the number of events
+  char eventstring[7];
```
Comment (matthiasrichter) [inline Generators/src/GeneratorFromFile.cxx]: one more candidate for the stringstream

**Comment by sawenzel** [comment]
Comments have been addressed. Merging.

File: Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h [hunk 10]
```diff
@@ -67,8 +67,11 @@ class DigitizerTask : public FairTask{
 
     int                 mTimeBinMax;   ///< Maximum time bin to be written out
     bool                mIsContinuousReadout; ///< Switch for continuous readout
+    int                 mHitSector=-1; ///< which sector to treat
 
-  ClassDefOverride(DigitizerTask, 1);
+    TClonesArray        *mSectorHitsArray[18];
```
Comment (ktf) [inline Detectors/TPC/simulation/include/TPCSimulation/DigitizerTask.h]: Maybe we should define and use a `TPC_MAX_SECTORS` constant in place of 18?

File: Detectors/TPC/simulation/src/Detector.cxx [hunk 11]
```diff
@@ -72,15 +85,28 @@ Detector::Detector(const char* name, Bool_t active)
...
     mPointCollection->Delete();
     delete mPointCollection;
   }
+#ifdef TPC_GROUPED_HITS
+  for(int i=0;i<18;++i){
+    mHitsPerSectorCollection[i]->Delete();
+    delete mHitsPerSectorCollection[i];
+  }
+#endif
+  std::cout << "Produced hits " << mHitCounter << "\n";
```
Comment (ktf) [inline Detectors/TPC/simulation/src/Detector.cxx]: Why not using LOG(INFO) or similar?

**Pull request merged**